sql/sqlcheck/sqlcheck.go

```diff
@@ -94,74 +94,118 @@ func (f ReportWriterFunc) WriteReport(r Report) {
 	f(r)
 }
 
-// Destructive checks destructive changes.
-var Destructive = &driverAware{
-	run: func(ctx context.Context, diags []Diagnostic, p *Pass) error {
-		for _, sc := range p.File.Changes {
-			for _, c := range sc.Changes {
-				switch c := c.(type) {
-				case *schema.DropSchema:
-					diags = append(diags, Diagnostic{
-						Pos:  sc.Pos,
-						Text: fmt.Sprintf("Dropping schema %q", c.S.Name),
-					})
-				case *schema.DropTable:
-					diags = append(diags, Diagnostic{
-						Pos:  sc.Pos,
-						Text: fmt.Sprintf("Dropping table %q", c.T.Name),
-					})
-				case *schema.ModifyTable:
-					for i := range c.Changes {
-						d, ok := c.Changes[i].(*schema.DropColumn)
-						if !ok {
-							continue
-						}
-						if g := (schema.GeneratedExpr{}); !sqlx.Has(d.C.Attrs, &g) || g.Type != "VIRTUAL" {
-							diags = append(diags, Diagnostic{
-								Pos:  sc.Pos,
-								Text: fmt.Sprintf("Dropping non-virtual column %q", d.C.Name),
-							})
-						}
+// ResourceSpan describes the lifespan of a resource
+// in perspective to the migration file.
+type ResourceSpan uint
+
+const (
+	// SpanUnknown describes unknown lifespan.
+	// e.g. resource may exist before this file.
+	SpanUnknown ResourceSpan = iota
+
+	// SpanAdded describes that a span of
+	// a resource was started in this file.
+	SpanAdded
+
+	// SpanDropped describes that a span of
+	// a resource was ended in this file.
+	SpanDropped
+
+	// SpanTemporary indicates that a resource lifetime
+	// was started and ended in this file (CREATE and DROP).
+	SpanTemporary = SpanAdded | SpanDropped
+)
```

Comment:
This may look overkill for one analyzer, but I also use it in the `DataDepend` analyzer (next PR).